packages/gatsby/src/commands/develop.ts

```diff
@@ -403,26 +404,10 @@ module.exports = async (program: IProgram): Promise<void> => {
     developProcess.send(msg)
   })
 
-  process.on(`beforeExit`, async () => {
-    await Promise.all([
-      watcher?.close(),
-      ...unlocks.map(unlock => unlock()),
-      new Promise(resolve => {
-        statusServer.close(resolve)
-      }),
-      new Promise(resolve => {
-        proxy.server.close(resolve)
-      }),
-    ])
-  })
-
-  process.on(`SIGINT`, async () => {
-    await developProcess.stop(`SIGINT`)
-    process.exit()
-  })
-
-  process.on(`SIGTERM`, async () => {
-    await developProcess.stop(`SIGTERM`)
-    process.exit()
+  onExit((code, signal) => {
+    developProcess.stop(signal)
+    unlocks.map(unlock => unlock())
+    statusServer.close()
+    proxy.server.close()
```

Comment:
Some of those are async, I'm particularly looking at [`server.close`](https://nodejs.org/api/net.html#net_server_close_callback). I'm not sure if this being `async`, but still working is something we can rely on fully.

`signal-exit` doesn't support async callbacks (understandably), but maybe we could put current `onbeforeexit` handler to it's own function (`shutdownServices()` ?) and call those in `SIGINT`, `SIGTERM`, await on it and only then call `process.exit` ? I mean - we could still keep the `signal-exit` to try to cover cases that we don't explicitly handle, but I think trying to shutdown gracefully on `SIGTERM` and actually awaiting on server instances closing would be possibly good idea? Potentially more resilient?

Reply:
I did some extensive testing and they all run the way they should. Problem is that none of the signals actually allow async await. Only onbeforeexit but it's rarely called

Reply:
Yes, but us handling `SIGINT` and `SIGTERM` allow us to await async code before calling `process.exit()`, right?

Reply:
True but this won't work with process.exit if I recall correctly. Fair enough i'll create a shutdown function for sigterm and sigint, still need to figure out process.exit

Reply:
Let's keep the code in `signal-exit` that you have (as last effort to close things if they weren't closed before).

I just think we can try to gracefully close things in some cases and not rely on `signal-exit` for those (which just doesn't support async stuff). So this is what I'm thinking about (pseudocode):

```js
async function shutdown() {
 // with Promise.all we currently have this will be for sure more complicated, but I hope it conveys what I mean
 if (somethingThatNeedsClosing) {
  await somethingThatNeedsClosing.close()
  somethingThatNeedsClosing = null
 }
}

process.on(`SIGTERM`, async () => {
  await shutdown()
  process.exit()
})

// last effort in case we didn't shutdown already (unhandled shutdown cases)
onExit(() => {
 if (somethingThatNeedsClosing) {
  somethingThatNeedsClosing.close()
  somethingThatNeedsClosing = null
 }
}
```